File: src/rtt_serial_cpp_pkg/src/rtt_serial_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <libserial/SerialStream.h>
#include <chrono>
#include <sstream>

using namespace LibSerial;
using namespace std::chrono;

class RTTSerialNode : public rclcpp::Node {
public:
    RTTSerialNode() : Node("rtt_serial_node") {
        try {
            serial_port_.Open("/dev/ttyUSB0");
            serial_port_.SetBaudRate(BaudRate::BAUD_115200);
            serial_port_.SetCharacterSize(CharacterSize::CHAR_SIZE_8);
            serial_port_.SetStopBits(StopBits::STOP_BITS_1);
            serial_port_.SetParity(Parity::PARITY_NONE);
            RCLCPP_INFO(this->get_logger(), "Serial port geöffnet.");
        } catch (const OpenFailed&) {
            RCLCPP_ERROR(this->get_logger(), "Fehler beim Öffnen des Ports!");
            rclcpp::shutdown();
        }

        timer_ = this->create_wall_timer(
            1000ms,
            std::bind(&RTTSerialNode::perform_rtt_test, this)
        );
    }

private:
    SerialStream serial_port_;
    rclcpp::TimerBase::SharedPtr timer_;

    void perform_rtt_test() {
        auto send_time = high_resolution_clock::now();
        std::ostringstream oss;
        oss << duration_cast<nanoseconds>(send_time.time_since_epoch()).count() << "\n";

        serial_port_ << oss.str();
        RCLCPP_INFO(this->get_logger(), "Gesendet: %s", oss.str().c_str());

        std::string line;
        std::getline(serial_port_, line);

        if (!line.empty()) {
            auto receive_time = high_resolution_clock::now();
            auto rtt = duration_cast<milliseconds>(receive_time - send_time).count();
            RCLCPP_INFO(this->get_logger(), "Antwort erhalten: %s | RTT: %ld micros", line.c_str(), rtt);
        } else {
            RCLCPP_WARN(this->get_logger(), "Keine Antwort erhalten.");
        }
    }
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<RTTSerialNode>());
    rclcpp::shutdown();
    return 0;
}

File: src/rtt_serial_pkg/rtt_serial_pkg/rtt_serial_node.py
import rclpy
from rclpy.node import Node
import serial
import time

class RTTSerialNode(Node):
    def __init__(self):
        super().__init__('rtt_serial_node')
        try:
            self.ser = serial.Serial('/dev/ttyUSB0', 115200, timeout=1)
            self.get_logger().info('Serial port geöffnet.')
        except serial.SerialException:
            self.get_logger().error('Fehler beim Öffnen des Ports!')
            rclpy.shutdown()
            return

        self.timer = self.create_timer(1.0, self.perform_rtt_test)

    def perform_rtt_test(self):
        send_time = time.time()
        msg = f"{send_time:.9f}\n"
        self.ser.write(msg.encode())
        self.get_logger().info(f'Gesendet: {msg.strip()}')

        try:
            response = self.ser.readline().decode().strip()
            if response:
                receive_time = time.time()
                rtt = (receive_time - send_time) * 1000
                self.get_logger().info(f'Antwort erhalten: {response} | RTT: {rtt:.2f} ms')
            else:
                self.get_logger().warn("Keine Antwort erhalten.")
        except Exception as e:
            self.get_logger().error(f'Fehler beim Lesen: {e}')

def main(args=None):
    rclpy.init(args=args)
    node = RTTSerialNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()